server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java

```diff
@@ -141,18 +162,24 @@ private DruidCoordinatorRuntimeParams runInternal(final DruidCoordinatorRuntimeP
     final CoordinatorRunStats stats = params.getCoordinatorStats();
 
     final int availableKillTaskSlots = getAvailableKillTaskSlots(dynamicConfig, stats);
-    Collection<String> dataSourcesToKill = dynamicConfig.getSpecificDataSourcesToKillUnusedSegmentsIn();
-
-    if (availableKillTaskSlots > 0) {
-      // If no datasource has been specified, all are eligible for killing unused segments
-      if (CollectionUtils.isNullOrEmpty(dataSourcesToKill)) {
-        dataSourcesToKill = segmentsMetadataManager.retrieveAllDataSourceNames();
-      }
+    if (availableKillTaskSlots <= 0) {
+      log.info("Skipping KillUnusedSegments because there are no available kill task slots.");
+      return params;
+    }
 
-      lastKillTime = DateTimes.nowUtc();
-      killUnusedSegments(dataSourcesToKill, availableKillTaskSlots, stats);
+    final Set<String> dataSourcesToKill;
+    if (!CollectionUtils.isNullOrEmpty(dynamicConfig.getSpecificDataSourcesToKillUnusedSegmentsIn())) {
+      dataSourcesToKill = dynamicConfig.getSpecificDataSourcesToKillUnusedSegmentsIn();
+    } else {
+      // If no datasource has been specified, all are eligible for killing unused segments by default
+      dataSourcesToKill = segmentsMetadataManager.retrieveAllDataSourceNames();
     }
 
+    datasourceIterator.updateCandidates(dataSourcesToKill);
```

Comment:
So this means that on every autoKill run cycle, we are affectively taking a new random start to the datasources to kill list, and working from there, right? If so, I think this is good.

Reply:
Also, if this is being updated each time, why keep the state at class level, instead of just having local variable that we use each time we go to kill unused segments. That way we dont have to worry about thread safety

Reply:
Yeah, exactly. If the datasources haven't changed, the iteration just resumes from the last position.